Tests/MapboxNavigationTests/NavigationMapViewTests.swift

```diff
@@ -390,34 +455,78 @@ class NavigationMapViewTests: TestCase {
         XCTAssertEqual(lineGradient[fractionTraveled], navigationMapView.trafficModerateColor)
         
         fractionTraveled = 0.45
         nextDownFractionTraveled = Double(CGFloat(fractionTraveled).nextDown)
-        lineGradient = navigationMapView.updateRouteLineGradientStops(fractionTraveled: fractionTraveled, gradientStops: lineGradient)
+        lineGradient = navigationMapView.updateRouteLineGradientStops(fractionTraveled: fractionTraveled, gradientStops: lineGradient, baseColor: navigationMapView.trafficUnknownColor)
         XCTAssertEqual(lineGradient[0.0], navigationMapView.traversedRouteColor)
         XCTAssertEqual(lineGradient[nextDownFractionTraveled], navigationMapView.traversedRouteColor)
         XCTAssertEqual(lineGradient[fractionTraveled], navigationMapView.trafficHeavyColor)
         
         lineGradient = [Double:UIColor]()
         fractionTraveled = 0.45
         nextDownFractionTraveled = Double(CGFloat(fractionTraveled).nextDown)
-        lineGradient = navigationMapView.updateRouteLineGradientStops(fractionTraveled: fractionTraveled, gradientStops: lineGradient)
+        lineGradient = navigationMapView.updateRouteLineGradientStops(fractionTraveled: fractionTraveled, gradientStops: lineGradient, baseColor: navigationMapView.trafficUnknownColor)
         XCTAssertEqual(lineGradient[0.0], navigationMapView.traversedRouteColor)
         XCTAssertEqual(lineGradient[nextDownFractionTraveled], navigationMapView.traversedRouteColor)
         XCTAssertEqual(lineGradient[fractionTraveled], navigationMapView.trafficUnknownColor)
     }
     
+    func testGenerateRouteLineRestrictedGradient() {
+        let coordinates: [CLLocationCoordinate2D] = [
+            CLLocationCoordinate2D(latitude: 1, longitude: 0),
+            CLLocationCoordinate2D(latitude: 2, longitude: 0),
+            CLLocationCoordinate2D(latitude: 3, longitude: 0),
+            CLLocationCoordinate2D(latitude: 4, longitude: 0),
+            CLLocationCoordinate2D(latitude: 5, longitude: 0),
+        ]
+        
+        let composeFeature = { (index: Int, isRestricted: Bool) -> Turf.Feature in
+            var feature = Feature.init(geometry: .lineString(LineString([coordinates[index],coordinates[index+1]])))
+            feature.properties = ["isRestrictedRoad": .boolean(isRestricted)]
+            return feature
+        }
+        
+        let features:[Turf.Feature] = [
+            composeFeature(0, true),
+            composeFeature(1, false),
+            composeFeature(2, true),
+            composeFeature(3, false)
+        ]
+        
+        var fractionTraveled = 0.0 // not traversed at all
+        var routeLineGradient = navigationMapView.routeLineRestrictionStops(features, fractionTraveled: fractionTraveled)
+        XCTAssertEqual(routeLineGradient[0.0], navigationMapView.routeRestrictedAreaColor)
+        XCTAssertEqual(routeLineGradient[1.0], navigationMapView.traversedRouteColor)
```

Comment:
After modifications of route line gradients calculation in [this commit](https://github.com/mapbox/mapbox-navigation-ios/commit/c949f78baabf84779b6a1bcbd99cc25fbe9e1db6) this test starts failing. I see that it has changed how ending segment is handled, but it still looks odd to me: In this particular case `routeLineGradient[0.75] == routeRestrictedAreaColor` (0.75 is the border between 2nd restricted segment and last non-restricted segment). Since it is "not soft" gradient, I would expect that `routeLineGradient[0.75.nextUp] == traversedColor`, but that is not the case and in fact it's `routeLineGradient[0.75.nextUp.nextUp] == traversedColor`. I am not sure why there is this little gap and how it appears. 
@ShanMa1991 maybe you could help me with it?

Reply:
Hi, yes, this gap exits. Because right now when we choose `soft`, we expected that the gradient stop of one feature is the `[segmentStartPercentTraveled + stopGap, segmentEndPercentTraveled - stopGap]`. Because the `stopGap` is based on the length of current congestion segment length, which would avoid the out of bound issue between a very long segment and a very short segment.

and when we choose `sharp` gradient, we actually still use the constraints of the current segment as `[segmentStartPercentTraveled.nextUp, segmentEndPercentTraveled.nextDown]`. So for example, for the 3rd segment in this test, the actual end percentage is `0.7500000000000001`, but we record the `0.75` for its end. And for the 4th segment, the actual start percentage is `0.7500000000000001`, because it's different than the 3rd one, we record it with `0.7500000000000002` as the start. Thus there's a `0.0000000000000002` gap between the 3rd segment and the 4th segment in the line gradient stops. It's larger than the previous `0.0000000000000001` but could avoid the check of next segment and the segment length issue.

And for the `0.0000000000000002` gap, because right now when we choose `sharp` gradient, we're using the following Expression:
```
        lineLayer.lineGradient = .expression(
            Exp(.step) {
                Exp(.lineProgress)
                UIColor.blue
                0.75
                UIColor.blue
                0.7500000000000002
                UIColor.red
            }
        )
```
And we have a following result, which still shows a sharp transition in the lowest zoom level.
![double](https://user-images.githubusercontent.com/48976398/143135539-fe8ad8f6-2f22-40b7-abc0-5bde5ae4e061.png)

And the edge case is that what if the `fractionTraveled` is in this gap, for example, `0.7500000000000001`.  In this case, the `minimumSegment` would record `(0.75, associatedFeatureColor)`, and then apply to the `gradientStops`. There's still a sharp change between the two segments.